File: packages/ts-host/src/a2ui.ts
import { CATALOG_ID } from '@ai37/a2ui-catalog-schemas'
import type { A2uiComponent, A2uiDataPatch } from './types'

/** Операция A2UI-поверхности (протокол v0.9: createSurface/updateComponents/updateDataModel/...). */
export type A2uiMessage = Record<string, unknown>

/** Плоская запись компонента в `updateComponents.components` (id + имя + props + слот-ссылки). */
type FlatComponent = Record<string, unknown> & { id: string; component: string }

/**
 * Уплощает дерево `A2uiComponent` в плоский список компонентов протокола v0.9: каждому узлу — `id`
 * (корень → `'root'`, детям — ДЕТЕРМИНИРОВАННЫЕ `${parentId}.${slot}[.${i}]`, важно для апсерта по id
 * при стриминге), в prop-слот родителя кладётся id-ссылка (строка для `ComponentIdSchema`, `string[]`
 * для `ChildListSchema`). Inline-вложенность протоколом запрещена.
 */
function flatten(node: A2uiComponent, id: string, out: FlatComponent[]): void {
  // Корень/узел добавляется первым; слот-props дописываются после раскрытия детей (id-ссылки).
  const entry: FlatComponent = { id, component: node.component, ...node.props }
  out.push(entry)

  if (!node.children) return
  for (const [slot, child] of Object.entries(node.children)) {
    if (Array.isArray(child)) {
      entry[slot] = child.map((c, i) => {
        const childId = c.id ?? `${id}.${slot}.${i}`
        flatten(c, childId, out)
        return childId
      })
    } else {
      const childId = child.id ?? `${id}.${slot}`
      flatten(child, childId, out)
      entry[slot] = childId
    }
  }
}

/**
 * Конвертирует декларативный компонент-дерево каталога в v0.9-операции A2UI-поверхности:
 * `createSurface` (с `catalogId`) + `updateComponents` с уплощённым деревом (корень `id: "root"`).
 * `catalogId`: явный `opts.catalogId` → тег компонента → дефолтный ai37 `CATALOG_ID`.
 *
 * Результат кладётся в `content.a2ui_operations` activity-сообщения `a2ui-surface`,
 * которое нативно рендерит CopilotKit v2 (`createA2UIMessageRenderer` + ai37Catalog).
 */
export function componentToA2uiOperations(
  component: A2uiComponent,
  opts: { surfaceId: string; catalogId?: string; dataModel?: A2uiDataPatch[] },
): A2uiMessage[] {
  const catalogId = opts.catalogId ?? component.catalogId ?? CATALOG_ID
  const components: FlatComponent[] = []
  // Корень surface всегда `'root'` (протокол v0.9); `component.id` верхнего узла игнорируется.
  flatten(component, 'root', components)
  return [
    { version: 'v0.9', createSurface: { surfaceId: opts.surfaceId, catalogId } },
    {
      version: 'v0.9',
      updateComponents: { surfaceId: opts.surfaceId, components },
    },
    // Патчи dataModel (напр. опции lookup-канала FormCard) — после компонентов,
    // чтобы подписчики читали значение с уже живого surface. `path` — точная
    // строка (ведущий слэш значим для клиента).
    ...(opts.dataModel ?? []).map((patch) => ({
      version: 'v0.9',
      updateDataModel: { surfaceId: opts.surfaceId, path: patch.path, value: patch.value },
    })),
  ]
}

File: packages/python-host/src/ai37_agent_host/a2ui.py
"""Уплощение дерева A2UI в v0.9-операции поверхности — порт ``ts-host/src/a2ui.ts``.

Протокол v0.9 не допускает inline-вложенность: дерево уплощается в плоский список
компонентов, дети — по детерминированным id-ссылкам (важно для апсерта по id при стриминге).
"""

from __future__ import annotations

from typing import Any

from .types import A2uiComponent

# ``CATALOG_ID`` (дефолтный ai37-каталог) живёт в python-пакете каталога, который может быть
# не установлен (другой репозиторий, не в индексе). Импорт мягкий: на практике create_agent_host
# всегда передаёт catalog_id явно из agent-card, дефолт нужен редко.
try:  # pragma: no cover - зависит от наличия пакета каталога
    from ai37_a2ui_catalog import CATALOG_ID as DEFAULT_CATALOG_ID
except Exception:  # pragma: no cover
    DEFAULT_CATALOG_ID = None

#: Операция A2UI-поверхности (v0.9: createSurface / updateComponents / ...).
A2uiMessage = dict[str, Any]


def _flatten(node: A2uiComponent, node_id: str, out: list[dict[str, Any]]) -> None:
    # Узел добавляется первым; слот-props (id-ссылки) дописываются после раскрытия детей.
    entry: dict[str, Any] = {"id": node_id, "component": node.component, **node.props}
    out.append(entry)

    if not node.children:
        return
    for slot, child in node.children.items():
        if isinstance(child, list):
            ids: list[str] = []
            for i, c in enumerate(child):
                child_id = c.id or f"{node_id}.{slot}.{i}"
                _flatten(c, child_id, out)
                ids.append(child_id)
            entry[slot] = ids
        else:
            child_id = child.id or f"{node_id}.{slot}"
            _flatten(child, child_id, out)
            entry[slot] = child_id


def component_to_a2ui_operations(
    component: A2uiComponent,
    *,
    surface_id: str,
    catalog_id: str | None = None,
) -> list[A2uiMessage]:
    """Дерево-компонент → ``createSurface`` (+catalogId) + ``updateComponents`` (уплощённое дерево).

    ``catalog_id``: явный аргумент → тег компонента → дефолтный ``CATALOG_ID``.
    Корень surface всегда ``'root'`` (v0.9); ``component.id`` верхнего узла игнорируется.
    """
    resolved = catalog_id or component.catalog_id or DEFAULT_CATALOG_ID
    components: list[dict[str, Any]] = []
    _flatten(component, "root", components)
    return [
        {
            "version": "v0.9",
            "createSurface": {"surfaceId": surface_id, "catalogId": resolved},
        },
        {
            "version": "v0.9",
            "updateComponents": {"surfaceId": surface_id, "components": components},
        },
    ]
